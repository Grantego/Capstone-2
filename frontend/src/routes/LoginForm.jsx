import React from "react";
import { Form, redirect, useNavigate, useActionData } from "react-router-dom";
import SpellerApi from "../../api";
import { FormGroup, Label, Input, Button } from "reactstrap";
import NavBar from "../components/NavBar";
import "./body.css"


export async function action({request}) {
    const errors = {};
    try {
        const formData = await request.formData();
        const data = Object.fromEntries(formData);
        const res = await SpellerApi.login(data);
        console.log(res)
        SpellerApi.token = res.token
        window.localStorage.setItem("username", data.username);
        window.localStorage.setItem("token", res.token);
        return redirect("/chords");
    } catch(e) {
        errors.unauthorized = "Username / Password incorrect!"
        return errors;
    }

}

const LoginForm = () => { 
    const errors = useActionData()

    let navigate = useNavigate()

    return (
        <>
            <NavBar/>
            <div className="body-container">
            <h1>Login</h1>
            {errors?.unauthorized && <span>{errors.unauthorized}</span>}
            <Form method="post">
                <FormGroup>
                    <Label htmlFor="username">Username:</Label>
                    <Input 
                        type="text"
                        name="username" 
                        id="username"
                        autoComplete="username"
                        required/>
                    <Label htmlFor="password">Password:</Label>
                    <Input 
                        type="password"
                        name="password" 
                        id="password"
                        autoComplete="current-password"
                        required/>
                </FormGroup>
                    <Button type="button" onClick={() => navigate(-1)}>Cancel</Button>
                    <Button type="submit">Login</Button>
            </Form>
            </div>
        </>
    )
}

export default LoginForm