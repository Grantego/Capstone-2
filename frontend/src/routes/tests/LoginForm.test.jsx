import React from 'react';
import { render, screen } from '@testing-library/react';
import LoginForm from '../LoginForm'
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { describe, it, beforeEach, expect} from 'vitest';


describe('LoginForm component', () => {
    let router

  beforeEach(() => {
    localStorage.clear();

    router = createMemoryRouter([
        {
            path: '/',
            element: <LoginForm/>,
        },
    ], {
        initialEntries: ['/'],
    }); 


  });

  it("matches snapshot", () => {
    const {asFragment} = render(<RouterProvider router={router} />)
    expect(asFragment()).toMatchSnapshot();
  })

  it("renders without crashing", () => {
    render(<RouterProvider router={router} />)
  })

  it("renders correct elements", () => {
    render(<RouterProvider router={router} />)
    expect(screen.getByPlaceholderText("enter username")).toBeInTheDocument()    
    expect(screen.getByPlaceholderText("enter password")).toBeInTheDocument()    
    expect(screen.getByLabelText("Username:")).toBeInTheDocument()    
    expect(screen.getByLabelText("Password:")).toBeInTheDocument()    
    expect(screen.getAllByText("Login")).toBeInTheDocument
  })

});