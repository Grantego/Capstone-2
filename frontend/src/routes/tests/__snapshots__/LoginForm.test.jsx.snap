// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChordDetails component > matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-light bg-faded"
  >
    <div
      class="container-fluid"
    >
      <a
        class="me-auto navbar-brand"
        href="/"
      >
        Chord Helper
      </a>
      <button
        aria-label="Toggle navigation"
        class="me-2 navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse"
      >
        <ul
          class="navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/signup"
            >
              Sign Up
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/login"
            >
              Login
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    class="body-container"
  >
    <h1>
      Login
    </h1>
    <form
      action="/"
      method="post"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="username"
        >
          Username:
        </label>
        <input
          autocomplete="username"
          class="form-control"
          id="username"
          name="username"
          placeholder="enter username"
          required=""
          type="text"
        />
        <label
          class="form-label"
          for="password"
        >
          Password:
        </label>
        <input
          autocomplete="current-password"
          class="form-control"
          id="password"
          name="password"
          placeholder="enter password"
          required=""
          type="password"
        />
      </div>
      <a
        class="btn btn-secondary"
        href="/"
        type="button"
      >
        Cancel
      </a>
      <button
        class="btn btn-secondary"
        id="login-btn"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`LoginForm component > matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-light bg-faded"
  >
    <div
      class="container-fluid"
    >
      <a
        class="me-auto navbar-brand"
        href="/"
      >
        Chord Helper
      </a>
      <button
        aria-label="Toggle navigation"
        class="me-2 navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse"
      >
        <ul
          class="navbar-nav"
        >
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/signup"
            >
              Sign Up
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class=""
              href="/login"
            >
              Login
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div
    class="body-container"
  >
    <h1>
      Login
    </h1>
    <form
      action="/"
      method="post"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="username"
        >
          Username:
        </label>
        <input
          autocomplete="username"
          class="form-control"
          id="username"
          name="username"
          placeholder="enter username"
          required=""
          type="text"
        />
        <label
          class="form-label"
          for="password"
        >
          Password:
        </label>
        <input
          autocomplete="current-password"
          class="form-control"
          id="password"
          name="password"
          placeholder="enter password"
          required=""
          type="password"
        />
      </div>
      <a
        class="btn btn-secondary"
        href="/"
        type="button"
      >
        Cancel
      </a>
      <button
        class="btn btn-secondary"
        id="login-btn"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</DocumentFragment>
`;
